File: AlgorithmicToolbox/week2/Assignment2/lastDigitSumOfFibonacciNumber/cpp/lastDigitSum.cpp
 #include <iostream>
using namespace std;
long long int repetation(long m)
{
    long long int prev = 0;
    long long int current = 1;
    long long int res = 0;
    for (long long int i = 0; i < m * m; i++)
    {
        long long int temp = (prev + current) % m;
        prev = current;
        current = temp;
        if (prev == 0 && current == 1)
            res = i + 1;
    }
    return res;
    }


    long long int fib(long long int n, long long int m)
    {
        if (n <= 1)
            return n;
        long long int curr = 1;
        long long int prev = 0;
        for (long long int i = 0; i < n - 1; i++)
        {
            long long int temp = (curr + prev) % m;
            prev = curr % m;
            curr = temp % m;
        }
        return curr;
    }

int main()
    {
        long long int n;
        cin >> n;
        long long int mod;
        mod = 10;
        if(n<=10){
            cout << fib(n,10)%10;
        }
        else{
            long long int temp = fib(n % repetation(10), 10) % 10;
            long long int q = n / repetation(10);
            cout<< ((temp + ((fib(repetation(10), 10) * q) % 10)) % 10);
        }
        return 0;
    }

File: AlgorithmicToolbox/week2/Assignment2/lastDigitSumOfFibonacciNumber/python/lastDigitSum.py
def repetation(m):
    prev = 0
    current = 1    
    res = 0
    for i in range(0,m*m):
        temp = (prev + current) % m
        prev = current
        current = temp
        if (prev == 0 and current == 1):
                res = i + 1
    return res
    

def fib(n, m):
    if (n <= 1):
        return n
    curr = 1
    prev = 0
    sum = 1
    for i in range(0,n-1):
        temp = (curr + prev) % m
        prev = curr % m
        curr = temp % m
        sum += curr % m
    return sum

    
n = int(input())

if (n <= 60):
    print(fib(n , 10)%10)
else:
    # print(n%repetation(10))
    temp = fib(n % repetation(10), 10) % 10
    # print(temp)
    q = (int)(n / repetation(10))
    # print(q)
    print( (temp + ((fib(repetation(10),10)*q)%10))%10)